server/lambda_function.py

```diff
@@ -0,0 +1,357 @@
+        if sub:
+            token['sub'] = sub
+        token['jti'] = str(uuid.uuid1())
+        token['iss'] = base_uri
+        token['aud'] = base_uri
+        token['exp'] = expires
+        token['iat'] = now
+        token['nbf'] = now
+        logger.debug('JWT params: %s', token)
+        return token
+
+    class Cookie:
+        def __init__(self):
+            logger.debug('Creating cookie')
+            self.cookie = cookies.SimpleCookie()
+
+        def remove_jwt(self, key):
+            logger.debug('Removing jwt %s', key)
+            self.cookie[key] = ''
+            self.cookie[key]['max-age'] = 0
+
+        def add_jwt(self, key, token, path='/', max_age=300, secure=True):
+            logger.debug(
+                'Adding jwt key: %s, path: %s, max age: %s, secure: %s',
+                 key,
+                 path,
+                 max_age,
+                 secure
+            )
+            self.cookie[key] = jwt.encode(
+                token,
+
+                # The first secret is the current signer.
+                jwt_secrets[0],
+                algorithm=algorithm
+            ).decode("utf-8")
+            self.cookie[key]['max-age'] = max_age
+            self.cookie[key]['secure'] = secure
+            self.cookie[key]['path'] = path
+
+        def get_headers(self):
+            headers = {}
+            for i, key in enumerate(self.cookie):
+
+                # hack around unique headers in api-gateway
+                set_cookie = list('SET-COOKIE')
+                if i > len(set_cookie):
+                    raise 'Can only set at most '+str(len(set_cookie))
+                set_cookie[i] = set_cookie[i].lower()
+                headers["".join(set_cookie)] = self.cookie[key].OutputString()
+            return headers
+
+    def redirect_with_cookie(url, cookie=None):
+        """
+        Redirect to url.
+        All keyword args will be signed into a JWT with the response.
+        """
+        logger.debug('Redirecting with cookie')
+        logger.debug('Redirecting to %s', url)
+        result = {
+            'statusCode': 302,
+            'headers': {
+                "Location": url
+            }
+        }
+        if cookie:
+            headers = result['headers'].copy()
+            headers.update(cookie.get_headers())
+            result['headers'] = headers
+
+        return result
+
+    def get_jwt_token(headers, key):
+        """
+        Retrieve the jwt token's as a key, value object
+        from the flask request cookies object.
+        If no token exists, returns an empty object.
+        """
+        logger.debug('Retrieving jwt token')
+        logger.debug('jwt token key: %s', key)
+        if 'cookie' not in map(lambda x: x.lower(), headers):
+            logger.info('no jwt token found for key %s', key)
+            return {}
+
+        cookie = cookies.SimpleCookie()
+        cookie.load(headers.get('cookie'))
+
+        if key in cookie:
```

Comment:
This is a little tough because the inner logic may return, but only if it succeeds in some iteration of the loop.
So, I still need the outer eventual return nothing to deal with falling through the loop without returning.

Reply:
I am fine with this